File: src/renderer/src/lib/blocks/palette/turtle.ts
import { Order } from '../generator'
import type { BlockDefinition } from '../registry'
import { DEFAULT_PEN_COLOUR, FIELD_COLOUR_TYPE } from '../colour-field'
import { pyString } from '../py'

/**
 * TURTLE GRAPHICS (#1013, epic #1007).
 * =============================================================================
 *
 * The first hour. `micropython/turtle.py` and the Turtle instrument (#1003) are
 * the closest thing Snakie has to Scratch's stage — a sprite that moves when you
 * tell it to, drawing as it goes, **with nothing wired up**. Paired with blocks
 * that is a complete first lesson: on a Pico, or in the simulator, on a
 * Chromebook, in a classroom with no hardware budget at all.
 *
 * WHY `import turtle` AND `turtle.forward(100)`, and not `from turtle import
 * forward`. Both are the module-level functions the issue asks for rather than
 * the `Turtle` class, and the second is shorter. The qualified call wins on
 * three counts:
 *
 *  1. **It says where `forward` comes from.** A program that also has hardware
 *     blocks in it is a file where a bare `forward(100)` is one name among many;
 *     `turtle.forward(100)` is self-describing, which is the property that makes
 *     generated code worth reading.
 *  2. **The import section doesn't grow.** One `import turtle` covers eighteen
 *     functions. A from-import would add a name per block type used, so the head
 *     of the file would visibly lengthen as the drawing got more interesting —
 *     the thing #1010's import manager exists to prevent.
 *  3. **It cannot collide with the learner's own names.** `forward` and `left`
 *     are exactly the words a child names their own function or variable. The
 *     generator guards variables against imported MODULE names (`names.ts`), so
 *     one `turtle` is one word to protect instead of eighteen.
 *
 * ONE BLOCK PER LIBRARY FUNCTION, which is why `pen up` and `pen down` are two
 * blocks rather than one with a dropdown — and why `turn right` and `turn left`
 * are too. It is what Scratch does, so the muscle memory transfers; it means the
 * block face tells you what will happen without reading a dropdown; and it keeps
 * every emitter a one-liner with nothing to get wrong. (#1012's hardware blocks
 * DO use dropdowns, because there a pin block and its twelve neighbours were
 * competing for one flyout. Here the pairs are the point.)
 *
 * THE TURTLE'S CONVENTIONS ARE NOT CPYTHON'S, deliberately (see `turtle.py`):
 * heading `0` points UP and increases CLOCKWISE, so `right(90)` faces east —
 * like turning right in real life. The block text says "turn right", never
 * "clockwise", because that is the sentence a child already knows.
 */

/** A number socket. The default that fills it lives in the toolbox shadow below. */
const num = (name: string): Record<string, unknown> => ({
  type: 'input_value',
  name,
  check: 'Number'
})

/** The shadow that puts `value` in `name`'s socket in the flyout. */
const shadowNum = (name: string, value: number): Record<string, unknown> => ({
  [name]: { shadow: { type: 'math_number', fields: { NUM: value } } }
})

/** `import turtle` — every block here needs it and nothing else. */
const NEEDS_TURTLE = [{ module: 'turtle' }] as const

/**
 * The instrument every block here belongs to, by its `instruments-registry.ts` id.
 *
 * Dragging any one of them opens the Turtle instrument, because the drawing IS
 * the output: a `forward 100` whose line goes into a panel nobody opened is
 * indistinguishable from a block that did nothing at all, and "nothing happened"
 * is the impression that loses a beginner for good.
 */
const TURTLE_INSTRUMENT = 'turtle'

/** A block that calls one turtle function with no arguments. */
const call = (
  type: string,
  message0: string,
  fn: string,
  tooltip: string,
  help: string
): BlockDefinition => ({
  type,
  category: 'turtle',
  instrument: TURTLE_INSTRUMENT,
  help,
  json: {
    message0,
    previousStatement: null,
    nextStatement: null,
    tooltip
  },
  imports: NEEDS_TURTLE,
  code: () => `turtle.${fn}()\n`
})

/** A block that calls one turtle function with a single number socket. */
const callWithNumber = (
  type: string,
  message0: string,
  fn: string,
  arg: string,
  fallback: number,
  tooltip: string,
  help: string
): BlockDefinition => ({
  type,
  category: 'turtle',
  instrument: TURTLE_INSTRUMENT,
  help,
  json: {
    message0,
    args0: [num(arg)],
    inputsInline: true,
    previousStatement: null,
    nextStatement: null,
    tooltip
  },
  imports: NEEDS_TURTLE,
  toolbox: { inputs: shadowNum(arg, fallback) },
  code: (block, gen) =>
    `turtle.${fn}(${gen.valueToCode(block, arg, Order.NONE) || String(fallback)})\n`
})

/** A value block reading one turtle accessor. */
const sensor = (
  type: string,
  message0: string,
  fn: string,
  tooltip: string
): BlockDefinition => ({
  type,
  category: 'turtle',
  instrument: TURTLE_INSTRUMENT,
  help: 'inst-turtle',
  json: {
    message0,
    output: 'Number',
    tooltip
  },
  imports: NEEDS_TURTLE,
  // FUNCTION_CALL, not ATOMIC: `turtle.xcor()` is a call, so an exponent or a
  // unary minus around it needs no brackets but `(turtle.xcor())[0]` would.
  code: () => [`turtle.${fn}()`, Order.FUNCTION_CALL]
})

export const TURTLE_BLOCKS: BlockDefinition[] = [
  // -------------------------------------------------------------------- movement
  callWithNumber(
    'snakie_turtle_forward',
    'move forward %1',
    'forward',
    'STEPS',
    100,
    'Move the turtle forwards along the way it is facing, drawing if the pen is down.',
    'inst-turtle'
  ),
  callWithNumber(
    'snakie_turtle_backward',
    'move backward %1',
    'backward',
    'STEPS',
    100,
    'Move the turtle backwards without turning it round.',
    'inst-turtle'
  ),
  callWithNumber(
    'snakie_turtle_right',
    'turn right %1 degrees',
    'right',
    'ANGLE',
    90,
    'Turn the turtle clockwise, on the spot. 90 is a quarter turn.',
    'inst-turtle'
  ),
  callWithNumber(
    'snakie_turtle_left',
    'turn left %1 degrees',
    'left',
    'ANGLE',
    90,
    'Turn the turtle anticlockwise, on the spot. 90 is a quarter turn.',
    'inst-turtle'
  ),
  {
    type: 'snakie_turtle_goto',
    category: 'turtle',
    instrument: TURTLE_INSTRUMENT,
    help: 'inst-turtle',
    json: {
      message0: 'go to x %1 y %2',
      args0: [num('X'), num('Y')],
      inputsInline: true,
      previousStatement: null,
      nextStatement: null,
      tooltip:
        'Jump straight to a point. (0, 0) is the middle of the screen; y counts upwards.'
    },
    imports: NEEDS_TURTLE,
    toolbox: { inputs: { ...shadowNum('X', 0), ...shadowNum('Y', 0) } },
    code: (block, gen) => {
      const x = gen.valueToCode(block, 'X', Order.NONE) || '0'
      const y = gen.valueToCode(block, 'Y', Order.NONE) || '0'
      return `turtle.goto(${x}, ${y})\n`
    }
  },
  callWithNumber(
    'snakie_turtle_setheading',
    'point in direction %1 degrees',
    'setheading',
    'ANGLE',
    0,
    'Face an absolute direction: 0 is up, 90 is right, 180 is down, 270 is left.',
    'inst-turtle'
  ),
  call(
    'snakie_turtle_home',
    'go home',
    'home',
    'Go back to the middle of the screen, facing up. The drawing stays.',
    'inst-turtle'
  ),

  // ------------------------------------------------------------------------- pen
  call(
    'snakie_turtle_penup',
    'pen up',
    'penup',
    'Lift the pen: the turtle moves without leaving a line.',
    'inst-turtle'
  ),
  call(
    'snakie_turtle_pendown',
    'pen down',
    'pendown',
    'Lower the pen: the turtle draws as it moves.',
    'inst-turtle'
  ),
  {
    type: 'snakie_turtle_pencolour',
    category: 'turtle',
    instrument: TURTLE_INSTRUMENT,
    help: 'inst-turtle',
    json: {
      message0: 'set pen colour to %1',
      args0: [{ type: FIELD_COLOUR_TYPE, name: 'COLOUR', colour: DEFAULT_PEN_COLOUR }],
      inputsInline: true,
      previousStatement: null,
      nextStatement: null,
      tooltip: 'Choose the colour of the lines the turtle draws from now on.'
    },
    imports: NEEDS_TURTLE,
    // The colour is a NAME in the code (`'red'`, not `'#ff0000'`) — see
    // `colour-field.ts`. Through `pyString` rather than hand-quoting, so the
    // whole generated file quotes its strings one way.
    code: (block) =>
      `turtle.pencolor(${pyString(String(block.getFieldValue('COLOUR') ?? DEFAULT_PEN_COLOUR))})\n`
  },
  callWithNumber(
    'snakie_turtle_pensize',
    'set pen size to %1',
    'pensize',
    'WIDTH',
    2,
    'How thick the turtle draws, in pixels.',
    'inst-turtle'
  ),

  // ---------------------------------------------------------------------- screen
  call(
    'snakie_turtle_clear',
    'clear the drawing',
    'clear',
    'Wipe the lines but leave the turtle exactly where it is.',
    'inst-turtle'
  ),
  call(
    'snakie_turtle_reset',
    'start again',
    'reset',
    'Wipe the drawing AND send the turtle home — a clean screen.',
    'inst-turtle'
  ),
  callWithNumber(
    'snakie_turtle_speed',
    'set speed to %1',
    'speed',
    'SPEED',
    6,
    'How fast the Turtle instrument animates the drawing: 1 is slow, 10 is quick.',
    'inst-turtle'
  ),
  call(
    'snakie_turtle_hide',
    'hide the turtle',
    'hideturtle',
    'Hide the turtle sprite, leaving just the drawing.',
    'inst-turtle'
  ),
  call(
    'snakie_turtle_show',
    'show the turtle',
    'showturtle',
    'Show the turtle sprite again.',
    'inst-turtle'
  ),

  // --------------------------------------------------------------------- sensing
  sensor('snakie_turtle_x', 'x position', 'xcor', "The turtle's distance right of the middle."),
  sensor('snakie_turtle_y', 'y position', 'ycor', "The turtle's distance above the middle."),
  sensor(
    'snakie_turtle_heading',
    'direction',
    'heading',
    'Which way the turtle is facing, in degrees: 0 is up, 90 is right.'
  )
]

File: micropython/turtle.py
"""Snakie Turtle — Logo-style turtle graphics as plain MicroPython functions.

Copy this file onto your MicroPython board (or just use it in the Simulator)
and ``import`` it. There is no separate Logo language and no special REPL:
turtle commands are ordinary Python function calls, so everything you already
know about variables, loops and functions applies straight away::

    from turtle import forward, right

    for _ in range(4):
        forward(50)
        right(90)

Like ``instruments.py`` this is a TELEMETRY library, not a driver: each call
that changes the turtle's state prints ONE ``SNK TURT ...`` line and the IDE's
Turtle instrument parses the broadcast serial stream to draw the picture — so
it works non-invasively, even inside a running program, and identically
whether the code is executing on real hardware or in the browser simulator.

Coordinate + heading conventions (deliberately NOT the same as CPython's
``turtle`` module — see the Snakie docs for why)::

    * heading 0 points UP (north/away from you); angles increase CLOCKWISE,
      so right(90) faces east — exactly like turning right in real life.
    * the origin (0, 0) is the CENTRE of the canvas; x increases right,
      y increases UP.

The telemetry protocol
-----------------------

One ``print()`` per state change, ASCII, space-delimited::

    SNK TURT POS  <x> <y> <heading>              # position/heading update
    SNK TURT LINE <x1> <y1> <x2> <y2> <colour> <width>  # a drawn segment
    SNK TURT PEN  <0|1>                          # pen up(0) / down(1)
    SNK TURT VIS  <0|1>                          # turtle hidden(0) / shown(1)
    SNK TURT CLEAR                               # wipe the canvas

``<colour>`` is whatever ``pencolor()`` was given (a CSS colour name or a
``#rrggbb`` hex code), with any spaces replaced by ``_`` so it stays one
token on the wire.
"""

SENTINEL = "SNK"

# Library version. Bump this on ANY change to this file — mirrors
# `instruments.py`'s convention so a future "board library outdated" check can
# reuse the same comparison.
__version__ = "0.1.0"

_DEFAULT_COLOUR = "black"
_DEFAULT_WIDTH = 2


def _fmt(n):
    """Format a coordinate/angle compactly: an int stays bare, a float rounds
    to 3 decimal places (plenty for a screen) so ``SNK TURT`` lines stay short.
    Pure + side-effect-free so it is unit-testable under CPython.
    """
    if isinstance(n, bool):
        return "1" if n else "0"
    if isinstance(n, int):
        return str(n)
    r = round(n, 3)
    if r == int(r):
        return str(int(r))
    return str(r)


def _colour_token(colour):
    """Encode a colour name/hex as a single ASCII token (spaces -> '_')."""
    return str(colour).replace(" ", "_")


class Turtle:
    """One turtle: tracks position/heading/pen state and emits ``SNK TURT``
    telemetry on every change. Most programs use the module-level functions
    (below), which drive a single shared :data:`_default` instance; construct
    your own :class:`Turtle` for more than one on screen at once.
    """

    def __init__(self):
        self.x = 0.0
        self.y = 0.0
        self.heading = 0.0  # compass: 0 = north/up, clockwise positive
        self.pen = True
        self.colour = _DEFAULT_COLOUR
        self.width = _DEFAULT_WIDTH
        self.visible = True
        self._speed = 6
        self._announce_pos()

    def _announce_pos(self):
        print("%s TURT POS %s %s %s" % (SENTINEL, _fmt(self.x), _fmt(self.y), _fmt(self.heading)))

    def _move_to(self, nx, ny):
        """Move to ``(nx, ny)``, drawing a line first if the pen is down."""
        if self.pen:
            print(
                "%s TURT LINE %s %s %s %s %s %s"
                % (
                    SENTINEL,
                    _fmt(self.x),
                    _fmt(self.y),
                    _fmt(nx),
                    _fmt(ny),
                    _colour_token(self.colour),
                    _fmt(self.width),
                )
            )
        self.x = nx
        self.y = ny
        self._announce_pos()

    # --- movement ----------------------------------------------------------

    def forward(self, distance):
        """Move ``distance`` units along the current heading, drawing if the
        pen is down."""
        import math

        rad = math.radians(self.heading)
        # Compass heading (0 = +y/north, clockwise positive):
        #   dx = sin(heading), dy = cos(heading)
        self._move_to(self.x + distance * math.sin(rad), self.y + distance * math.cos(rad))

    def backward(self, distance):
        """Move ``distance`` units opposite the current heading, without
        turning."""
        self.forward(-distance)

    def right(self, angle):
        """Rotate clockwise by ``angle`` degrees."""
        self.heading = (self.heading + angle) % 360
        self._announce_pos()

    def left(self, angle):
        """Rotate anticlockwise by ``angle`` degrees."""
        self.right(-angle)

    # --- pen -----------------------------------------------------------

    def penup(self):
        """Lift the pen — further movement leaves no trail."""
        self.pen = False
        print("%s TURT PEN 0" % SENTINEL)

    def pendown(self):
        """Lower the pen — further movement draws a line."""
        self.pen = True
        print("%s TURT PEN 1" % SENTINEL)

    def pencolor(self, colour):
        """Set the line colour (a CSS colour name or ``#rrggbb`` hex)."""
        self.colour = colour

    def pensize(self, width):
        """Set the line width in pixels."""
        self.width = width

    # --- state and positioning ------------------------------------------

    def home(self):
        """Return to the centre, at the starting (north) heading."""
        self.goto(0, 0)
        self.setheading(0)

    def goto(self, x, y):
        """Jump to the absolute coordinate ``(x, y)``, drawing if pen is down."""
        self._move_to(float(x), float(y))

    def setheading(self, angle):
        """Point in an absolute compass direction (0 = north, clockwise)."""
        self.heading = float(angle) % 360
        self._announce_pos()

    def clear(self):
        """Wipe drawn lines, leaving the turtle's position/heading as-is."""
        print("%s TURT CLEAR" % SENTINEL)

    def reset(self):
        """Clear the canvas and return home, in one call."""
        self.clear()
        self.home()

    def speed(self, value=None):
        """Get/set the animation pace (a small int, or a CPython-turtle-style
        name like ``"fast"``/``"slow"``); the IDE's Turtle instrument uses this
        to pace its draw animation. Returns the current speed when called with
        no argument."""
        if value is None:
            return self._speed
        self._speed = value
        return None

    def hideturtle(self):
        """Hide the turtle sprite (handy for a finished drawing)."""
        self.visible = False
        print("%s TURT VIS 0" % SENTINEL)

    def showturtle(self):
        """Show the turtle sprite."""
        self.visible = True
        print("%s TURT VIS 1" % SENTINEL)

    def position(self):
        """Return the current ``(x, y)``."""
        return (self.x, self.y)


# ---------------------------------------------------------------------------
# Module-level API — the shared default turtle most programs use directly, so
# `forward(10)` / `right(90)` just work without instantiating a Turtle.
# ---------------------------------------------------------------------------

_default = Turtle()


def forward(distance):
    _default.forward(distance)


def backward(distance):
    _default.backward(distance)


def right(angle):
    _default.right(angle)


def left(angle):
    _default.left(angle)


def penup():
    _default.penup()


def pendown():
    _default.pendown()


def pencolor(colour):
    _default.pencolor(colour)


def pensize(width):
    _default.pensize(width)


def home():
    _default.home()


def goto(x, y):
    _default.goto(x, y)


def setheading(angle):
    _default.setheading(angle)


def clear():
    _default.clear()


def reset():
    _default.reset()


def speed(value=None):
    return _default.speed(value)


def hideturtle():
    _default.hideturtle()


def showturtle():
    _default.showturtle()


def position():
    return _default.position()


# Short aliases, kept for anyone porting CPython `turtle` code (`fd`/`bk`
# etc. are its own conventional short names, unrelated to classic Logo).
fd = forward
bk = backward
back = backward
rt = right
lt = left
pu = penup
pd = pendown
